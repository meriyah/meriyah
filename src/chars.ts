/**
 * A list of character constants with much more human-readable names.
 */
export const enum Chars {

    Null = 0x00,

    Backspace      = 0x08,
    Tab            = 0x09,
    LineFeed       = 0x0A,
    VerticalTab    = 0x0B,
    FormFeed       = 0x0C,
    CarriageReturn = 0x0D,

    Space       = 0x20,
    Exclamation = 0x21,
    DoubleQuote = 0x22,
    Hash        = 0x23,
    Dollar      = 0x24,
    Percent     = 0x25,
    Ampersand   = 0x26,
    SingleQuote = 0x27,
    LeftParen   = 0x28,
    RightParen  = 0x29,
    Asterisk    = 0x2A,
    Plus        = 0x2B,
    Comma       = 0x2C,
    Hyphen      = 0x2D,
    Period      = 0x2E,
    Slash       = 0x2F,

    Zero         = 0x30,
    One          = 0x31,
    Two          = 0x32,
    Three        = 0x33,
    Four         = 0x34,
    Five         = 0x35,
    Six          = 0x36,
    Seven        = 0x37,
    Eight        = 0x38,
    Nine         = 0x39,
    Colon        = 0x3A,
    Semicolon    = 0x3B,
    LessThan     = 0x3C,
    EqualSign    = 0x3D,
    GreaterThan  = 0x3E,
    QuestionMark = 0x3F,

    UpperA = 0x41,
    UpperB = 0x42,
    UpperC = 0x43,
    UpperD = 0x44,
    UpperE = 0x45,
    UpperF = 0x46,
    UpperG = 0x47,
    UpperH = 0x48,
    UpperI = 0x49,
    UpperJ = 0x4A,
    UpperK = 0x4B,
    UpperL = 0x4C,
    UpperM = 0x4D,
    UpperN = 0x4E,
    UpperO = 0x4F,

    UpperP       = 0x50,
    UpperQ       = 0x51,
    UpperR       = 0x52,
    UpperS       = 0x53,
    UpperT       = 0x54,
    UpperU       = 0x55,
    UpperV       = 0x56,
    UpperW       = 0x57,
    UpperX       = 0x58,
    UpperY       = 0x59,
    UpperZ       = 0x5A,
    LeftBracket  = 0x5B,
    Backslash    = 0x5C,
    RightBracket = 0x5D,
    Caret        = 0x5E,
    Underscore   = 0x5F,

    Backtick = 0x60,
    LowerA   = 0x61,
    LowerB   = 0x62,
    LowerC   = 0x63,
    LowerD   = 0x64,
    LowerE   = 0x65,
    LowerF   = 0x66,
    LowerG   = 0x67,
    LowerH   = 0x68,
    LowerI   = 0x69,
    LowerJ   = 0x6A,
    LowerK   = 0x6B,
    LowerL   = 0x6C,
    LowerM   = 0x6D,
    LowerN   = 0x6E,
    LowerO   = 0x6F,

    LowerP      = 0x70,
    LowerQ      = 0x71,
    LowerR      = 0x72,
    LowerS      = 0x73,
    LowerT      = 0x74,
    LowerU      = 0x75,
    LowerV      = 0x76,
    LowerW      = 0x77,
    LowerX      = 0x78,
    LowerY      = 0x79,
    LowerZ      = 0x7A,
    LeftBrace   = 0x7B,
    VerticalBar = 0x7C,
    RightBrace  = 0x7D,
    Tilde       = 0x7E,
    Delete      = 0x7F,

    NextLine         = 0x85,
    NonBreakingSpace = 0xA0,

    Ogham = 0x1680,

    EnQuad             = 0x2000,
    EmQuad             = 0x2001,
    EnSpace            = 0x2002,
    EmSpace            = 0x2003,
    ThreePerEmSpace    = 0x2004,
    FourPerEmSpace     = 0x2005,
    SixPerEmSpace      = 0x2006,
    FigureSpace        = 0x2007,
    PunctuationSpace   = 0x2008,
    ThinSpace          = 0x2009,
    HairSpace          = 0x200A,
    ZeroWidthSpace     = 0x200B,
    ZeroWidthJoiner    = 0x200C,
    ZeroWidthNonJoiner = 0x200D,

    LineSeparator      = 0x2028,
    ParagraphSeparator = 0x2029,

    NarrowNoBreakSpace = 0x202F,
    MathematicalSpace  = 0x205F,
    IdeographicSpace   = 0x3000,

    ZeroWidthNoBreakSpace = 0xFEFF,

    ByteOrderMark = 0xFFEF,

    NonBMPMax = 0x10FFFF,

    LeadSurrogateMin = 0xD800,
    LeadSurrogateMax = 0xDBFF,
    TrailSurrogateMin = 0xDC00,
    TrailSurrogateMax = 0xDFFF,
    UTF16Max = 0xFFFF,
    NonBMPMin = 0x10000,
}
