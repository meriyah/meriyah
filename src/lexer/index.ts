export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
