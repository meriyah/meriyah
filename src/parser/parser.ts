import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
