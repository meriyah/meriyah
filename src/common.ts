import { Errors } from './errors';
import { nextToken } from './lexer/scan';
import { type Parser } from './parser/parser';
import { KeywordDescTable, Token } from './token';

/**
 * The core context, passed around everywhere as a simple immutable bit set
 */
export const enum Context {
  None = 0,
  Strict = 1 << 0,
  Module = 1 << 1, // Current code should be parsed as a module body
  InSwitch = 1 << 2,
  InGlobal = 1 << 3,
  InClass = 1 << 4,
  AllowRegExp = 1 << 5,
  TaggedTemplate = 1 << 6,
  InIteration = 1 << 7,
  SuperProperty = 1 << 8,
  SuperCall = 1 << 9,
  InYieldContext = 1 << 10,
  InAwaitContext = 1 << 11,
  InReturnContext = 1 << 12,
  InArgumentList = 1 << 13,
  InConstructor = 1 << 14,
  InMethodOrFunction = 1 << 15,
  AllowNewTarget = 1 << 16,
  DisallowIn = 1 << 17,
  AllowEscapedKeyword = 1 << 18,
  InStaticBlock = 1 << 19,
}

/**
 * Masks to track the property kind
 */
export const enum PropertyKind {
  None = 0,
  Method = 1 << 0,
  Computed = 1 << 1,
  Shorthand = 1 << 2,
  Generator = 1 << 3,
  Async = 1 << 4,
  Static = 1 << 5,
  Constructor = 1 << 6,
  ClassField = 1 << 7,
  Getter = 1 << 8,
  Setter = 1 << 9,
  Accessor = 1 << 10,
  Extends = 1 << 11,
  Literal = 1 << 12,
  PrivateField = 1 << 13,
  GetSet = Getter | Setter,
}

/**
 * Masks to track the binding kind
 */
export const enum BindingKind {
  None = 0,
  ArgumentList = 1 << 0,
  Empty = 1 << 1,
  Variable = 1 << 2,
  Let = 1 << 3,
  Const = 1 << 4,
  Class = 1 << 5,
  FunctionLexical = 1 << 6,
  FunctionStatement = 1 << 7,
  CatchPattern = 1 << 8,
  CatchIdentifier = 1 << 9,
  Async = 1 << 10,
  Generator = 1 << 10,
  AsyncFunctionLexical = Async | FunctionLexical,
  GeneratorFunctionLexical = Generator | FunctionLexical,
  AsyncGeneratorFunctionLexical = Async | Generator | FunctionLexical,
  CatchIdentifierOrPattern = CatchIdentifier | CatchPattern,
  LexicalOrFunction = Variable | FunctionLexical,
  LexicalBinding = Let | Const | FunctionLexical | FunctionStatement | Class,
}

/**
 * The masks to track where something begins. E.g. statements, declarations or arrows.
 */
export const enum Origin {
  None = 0,
  Statement = 1 << 0,
  BlockStatement = 1 << 1,
  TopLevel = 1 << 2,
  Declaration = 1 << 3,
  Arrow = 1 << 4,
  ForStatement = 1 << 5,
  Export = 1 << 6,
}

/**
 * Masks to track the assignment kind
 */
export const enum AssignmentKind {
  None = 0,
  Assignable = 1 << 0,
  CannotAssign = 1 << 1,
}

/**
 * Masks to track the destructuring kind
 */
export const enum DestructuringKind {
  None = 0,
  HasToDestruct = 1 << 3,
  // "Cannot" rather than "can" so that this flag can be ORed together across
  // multiple characters.
  CannotDestruct = 1 << 4,
  // Only destructible if assignable
  Assignable = 1 << 5,
  // `__proto__` is a special case and only valid to parse if destructible
  SeenProto = 1 << 6,
  Await = 1 << 7,
  Yield = 1 << 8,
}

/**
 * The mutable parser flags, in case any flags need passed by reference.
 */
export const enum Flags {
  None = 0,
  NewLine = 1 << 0,
  HasConstructor = 1 << 5,
  Octal = 1 << 6,
  NonSimpleParameterList = 1 << 7,
  HasStrictReserved = 1 << 8,
  StrictEvalArguments = 1 << 9,
  DisallowCall = 1 << 10,
  HasOptionalChaining = 1 << 11,
  EightAndNine = 1 << 12,
}

export const enum HoistedClassFlags {
  None,
  Hoisted = 1 << 0,
  Export = 1 << 1,
}

export const enum HoistedFunctionFlags {
  None,
  Hoisted = 1 << 0,
  Export = 1 << 1,
}

/**
 * Check for automatic semicolon insertion according to the rules
 * given in `ECMA-262, section 11.9`.
 *
 * @param parser Parser object
 * @param context Context masks
 */

export function matchOrInsertSemicolon(parser: Parser, context: Context): void {
  if ((parser.flags & Flags.NewLine) === 0 && (parser.getToken() & Token.IsAutoSemicolon) !== Token.IsAutoSemicolon) {
    parser.report(Errors.UnexpectedToken, KeywordDescTable[parser.getToken() & Token.Type]);
  }

  if (!consumeOpt(parser, context, Token.Semicolon)) {
    // Automatic semicolon insertion has occurred
    parser.options.onInsertedSemicolon?.(parser.startIndex);
  }
}

export function isValidStrictMode(parser: Parser, index: number, tokenIndex: number, tokenValue: string): 0 | 1 {
  if (index - tokenIndex < 13 && tokenValue === 'use strict') {
    if ((parser.getToken() & Token.IsAutoSemicolon) === Token.IsAutoSemicolon || parser.flags & Flags.NewLine) {
      return 1;
    }
  }
  return 0;
}

/**
 * Consumes the current token if the current token kind is
 * the specified `kind` and returns `0`. Otherwise returns `1`.
 *
 * @param parser Parser state
 * @param context Context masks
 * @param token The type of token to consume
 */
export function optionalBit(parser: Parser, context: Context, t: Token): 0 | 1 {
  if (parser.getToken() !== t) return 0;
  nextToken(parser, context);
  return 1;
}

/** Consumes the current token if the current token kind is
 * the specified `kind` and returns `true`. Otherwise returns
 * `false`.
 *
 * @param parser Parser state
 * @param context Context masks
 * @param token The type of token to consume
 */
export function consumeOpt(parser: Parser, context: Context, t: Token): boolean {
  if (parser.getToken() !== t) return false;
  nextToken(parser, context);
  return true;
}

/**
 * Consumes the current token. If the current token kind is not
 * the specified `kind`, an error will be reported.
 *
 * @param parser Parser state
 * @param context Context masks
 * @param t The type of token to consume
 */
export function consume(parser: Parser, context: Context, t: Token): void {
  if (parser.getToken() !== t) parser.report(Errors.ExpectedToken, KeywordDescTable[t & Token.Type]);
  nextToken(parser, context);
}

/**
 * Transforms a `LeftHandSideExpression` into a `AssignmentPattern` if possible,
 * otherwise it returns the original tree.
 *
 * @param parser Parser state
 * @param {*} node
 */
export function reinterpretToPattern(parser: Parser, node: any): void {
  switch (node.type) {
    case 'ArrayExpression': {
      node.type = 'ArrayPattern';
      const { elements } = node;
      for (let i = 0, n = elements.length; i < n; ++i) {
        const element = elements[i];
        if (element) reinterpretToPattern(parser, element);
      }
      return;
    }
    case 'ObjectExpression': {
      node.type = 'ObjectPattern';
      const { properties } = node;
      for (let i = 0, n = properties.length; i < n; ++i) {
        reinterpretToPattern(parser, properties[i]);
      }
      return;
    }
    case 'AssignmentExpression':
      node.type = 'AssignmentPattern';
      if (node.operator !== '=') parser.report(Errors.InvalidDestructuringTarget);
      delete node.operator;
      reinterpretToPattern(parser, node.left);
      return;
    case 'Property':
      reinterpretToPattern(parser, node.value);
      return;
    case 'SpreadElement':
      node.type = 'RestElement';
      reinterpretToPattern(parser, node.argument);
    // No default
  }
}

/**
 * Validates binding identifier
 *
 * @param parser Parser state
 * @param context Context masks
 * @param type Binding type
 * @param token Token
 */

export function validateBindingIdentifier(
  parser: Parser,
  context: Context,
  kind: BindingKind,
  t: Token,
  skipEvalArgCheck: 0 | 1,
): void {
  if (context & Context.Strict) {
    if ((t & Token.FutureReserved) === Token.FutureReserved) {
      parser.report(Errors.UnexpectedStrictReserved);
    }

    if (!skipEvalArgCheck && (t & Token.IsEvalOrArguments) === Token.IsEvalOrArguments) {
      parser.report(Errors.StrictEvalArguments);
    }
  }

  if ((t & Token.Reserved) === Token.Reserved || t === Token.EscapedReserved) {
    parser.report(Errors.KeywordNotId);
  }

  // The BoundNames of LexicalDeclaration and ForDeclaration must not
  // contain 'let'. (CatchParameter is the only lexical binding form
  // without this restriction.)
  if (kind & (BindingKind.Let | BindingKind.Const) && (t & Token.Type) === (Token.LetKeyword & Token.Type)) {
    parser.report(Errors.InvalidLetConstBinding);
  }

  if (context & (Context.InAwaitContext | Context.Module) && t === Token.AwaitKeyword) {
    parser.report(Errors.AwaitIdentInModuleOrAsyncFunc);
  }

  if (context & (Context.InYieldContext | Context.Strict) && t === Token.YieldKeyword) {
    parser.report(Errors.DisallowedInContext, 'yield');
  }
}

export function validateFunctionName(parser: Parser, context: Context, t: Token): void {
  if (context & Context.Strict) {
    if ((t & Token.FutureReserved) === Token.FutureReserved) {
      parser.report(Errors.UnexpectedStrictReserved);
    }

    if ((t & Token.IsEvalOrArguments) === Token.IsEvalOrArguments) {
      parser.report(Errors.StrictEvalArguments);
    }

    if (t === Token.EscapedFutureReserved) {
      parser.report(Errors.InvalidEscapedKeyword);
    }

    if (t === Token.EscapedReserved) {
      parser.report(Errors.InvalidEscapedKeyword);
    }
  }

  if ((t & Token.Reserved) === Token.Reserved) {
    parser.report(Errors.KeywordNotId);
  }

  if (context & (Context.InAwaitContext | Context.Module) && t === Token.AwaitKeyword) {
    parser.report(Errors.AwaitIdentInModuleOrAsyncFunc);
  }

  if (context & (Context.InYieldContext | Context.Strict) && t === Token.YieldKeyword) {
    parser.report(Errors.DisallowedInContext, 'yield');
  }
}

/**
 * Validates binding identifier
 *
 * @param parser Parser state
 * @param context Context masks
 * @param t Token
 */

export function isStrictReservedWord(parser: Parser, context: Context, t: Token): boolean {
  if (t === Token.AwaitKeyword) {
    if (context & (Context.InAwaitContext | Context.Module)) parser.report(Errors.AwaitIdentInModuleOrAsyncFunc);
    parser.destructible |= DestructuringKind.Await;
  }

  if (t === Token.YieldKeyword && context & Context.InYieldContext) parser.report(Errors.DisallowedInContext, 'yield');

  return (
    (t & Token.Reserved) === Token.Reserved ||
    (t & Token.FutureReserved) === Token.FutureReserved ||
    t == Token.EscapedFutureReserved
  );
}

/**
 * Checks if the property has any private field key
 *
 * @param parser Parser object
 * @param context  Context masks
 */
export function isPropertyWithPrivateFieldKey(expr: any): boolean {
  return !expr.property ? false : expr.property.type === 'PrivateIdentifier';
}

/**
 * Checks if a label in `LabelledStatement` are valid or not
 *
 * @param parser Parser state
 * @param labels Object holding the labels
 * @param name Current label
 * @param isIterationStatement
 */
export function isValidLabel(parser: Parser, labels: any, name: string, isIterationStatement: 0 | 1): 0 | 1 {
  while (labels) {
    if (labels['$' + name]) {
      if (isIterationStatement) parser.report(Errors.InvalidNestedStatement);
      return 1;
    }
    if (isIterationStatement && labels.loop) isIterationStatement = 0;
    labels = labels['$'];
  }

  return 0;
}

/**
 * Checks if current label already have been declared, and if not
 * declare it
 *
 * @param parser Parser state
 * @param labels Object holding the labels
 * @param name Current label
 */
export function validateAndDeclareLabel(parser: Parser, labels: any, name: string): void {
  let set = labels;
  while (set) {
    if (set['$' + name]) parser.report(Errors.LabelRedeclaration, name);
    set = set['$'];
  }

  labels['$' + name] = 1;
}

/** @internal */
export function isEqualTagName(elementName: any): any {
  switch (elementName.type) {
    case 'JSXIdentifier':
      return elementName.name;
    case 'JSXNamespacedName':
      return elementName.namespace + ':' + elementName.name;
    case 'JSXMemberExpression':
      return isEqualTagName(elementName.object) + '.' + isEqualTagName(elementName.property);
    /* istanbul ignore next */
    default:
    // ignore
  }
}

export function isValidIdentifier(context: Context, t: Token): boolean {
  if (context & (Context.Strict | Context.InYieldContext)) {
    // Module code is also "strict mode code"
    if (context & Context.Module && t === Token.AwaitKeyword) return false;
    if (context & Context.InYieldContext && t === Token.YieldKeyword) return false;
    return (t & Token.Contextual) === Token.Contextual;
  }

  return (t & Token.Contextual) === Token.Contextual || (t & Token.FutureReserved) === Token.FutureReserved;
}

export function classifyIdentifier(parser: Parser, context: Context, t: Token): any {
  if ((t & Token.IsEvalOrArguments) === Token.IsEvalOrArguments) {
    if (context & Context.Strict) parser.report(Errors.StrictEvalArguments);
    parser.flags |= Flags.StrictEvalArguments;
  }

  if (!isValidIdentifier(context, t)) parser.report(Errors.Unexpected);
}

export type Location = {
  readonly index: number;
  readonly line: number;
  readonly column: number;
};

export function getOwnProperty<T>(object: Record<string | number, T>, key: string | number): T | undefined {
  return Object.hasOwn(object, key) ? object[key] : undefined;
}
