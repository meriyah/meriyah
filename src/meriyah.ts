import { Context } from './common';
import { parseSource, Options } from './parser';
import * as ESTree from './estree';
// Current version
import { version } from '../package.json';

/**
 * Parse a script, optionally with various options.
 */
export function parseScript(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

/**
 * Parse a module, optionally with various options.
 */
export function parseModule(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.Strict | Context.Module);
}

/**
 * Parse a module or a script, optionally with various options.
 */
export function parse(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

export { Options, ESTree, version };
