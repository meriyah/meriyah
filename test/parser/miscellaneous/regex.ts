import * as t from 'node:assert/strict';
import { outdent } from 'outdent';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { fail } from '../../test-utils';

describe('Miscellaneous - Regular expressions', () => {
  for (const arg of [
    //    '/(?<abc𝟐def>foo\\k<abc𝟐def>)/',
    //    '/(?<輸xyz>foo)met\\k<輸xyz>/',
    String.raw`x = /[^\x00-\x2f\x3a-\x40\x5b-\x60\x7b-\x7f]+/g;`,
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { next: true, lexical: true });
      });
    });
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, { next: true, webcompat: true, lexical: true });
      });
    });
  }

  fail('Miscellaneous - Regular expressions (fail)', [
    { code: 'function *f(){   s = {foo: yield / x}   }', options: { next: true } },
    { code: 's = {foo: yield / x}', options: { impliedStrict: true, next: true } },
    { code: 'function *f(){   s = {"foo": yield / x}   }', options: { next: true } },
    { code: String.raw`/(?<a>.)\k<a/`, options: { next: true } },
    { code: String.raw`/\k<a(?<a>a)/`, options: { next: true } },
    { code: '/(?<42a>a)/', options: { next: true } },
    { code: '/(?<𝟐rest>foo)/', options: { next: true } },
    { code: '/(?<𝟐>foo)/', options: { next: true } },
    { code: String.raw`/(?<\uD835\uDFD0rest>foo)/`, options: { webcompat: true } },
    { code: String.raw`/(?<abc\uD835\uDFD0def>foo\k<abc\uD835def>)/`, options: { sourceType: 'module', next: true } },
    // Nodejs v18 now accepts unicode in capture group name
    // ['/(?<\\ud87e\\udddfrest>foo)/', Context.OptionsNext | Context.OptionsWebCompat],
    {
      code: outdent`
        function* f(){ yield
        /foo }
      `,
      options: { sourceType: 'module', next: true },
    },
    { code: 'function l(){((/)/))(/]/)};', options: { sourceType: 'module', next: true } },
    { code: '0 ?? 1 && 2', options: { sourceType: 'module', next: true } },
    {
      code: '3 ?? 2 ** 1 % 0 / 9 * 8 - 7 + 6 >>> 5 >> 4 << 3 >= 2 <= 1 > 0 < 9 !== 8 === 7 != 6 == 5 & 4 ^ 3 | 2 && 1 || 0',
      options: { next: true },
    },
    { code: 'e ?? f ?? g || h;', options: { sourceType: 'module', next: true } },
    { code: 'c && d ?? e', options: { sourceType: 'module', next: true } },
  ]);
});
