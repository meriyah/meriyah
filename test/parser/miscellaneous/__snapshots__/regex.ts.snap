// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > /(?<\\uD835\\uDFD0rest>foo)/ 1`] = `
"SyntaxError [1:0-1:26]: Unterminated regular expression
> 1 | /(?<\\uD835\\uDFD0rest>foo)/
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > /(?<42a>a)/ 1`] = `
"SyntaxError [1:0-1:11]: Unterminated regular expression
> 1 | /(?<42a>a)/
    | ^^^^^^^^^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > /(?<a>.)\\k<a/ 1`] = `
"SyntaxError [1:0-1:13]: Unterminated regular expression
> 1 | /(?<a>.)\\k<a/
    | ^^^^^^^^^^^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > /(?<abc\\uD835\\uDFD0def>foo\\k<abc\\uD835def>)/ 1`] = `
"SyntaxError [1:0-1:44]: Unterminated regular expression
> 1 | /(?<abc\\uD835\\uDFD0def>foo\\k<abc\\uD835def>)/
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > /(?<𝟐>foo)/ 1`] = `
"SyntaxError [1:0-1:12]: Unterminated regular expression
> 1 | /(?<𝟐>foo)/
    | ^^^^^^^^^^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > /(?<𝟐rest>foo)/ 1`] = `
"SyntaxError [1:0-1:16]: Unterminated regular expression
> 1 | /(?<𝟐rest>foo)/
    | ^^^^^^^^^^^^^^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > /\\k<a(?<a>a)/ 1`] = `
"SyntaxError [1:0-1:13]: Unterminated regular expression
> 1 | /\\k<a(?<a>a)/
    | ^^^^^^^^^^^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > 0 ?? 1 && 2 1`] = `
"SyntaxError [1:7-1:9]: Coalescing and logical operators used together in the same expression must be disambiguated with parentheses
> 1 | 0 ?? 1 && 2
    |        ^^ Coalescing and logical operators used together in the same expression must be disambiguated with parentheses"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > 3 ?? 2 ** 1 % 0 / 9 * 8 - 7 + 6 >>> 5 >> 4 << 3 >= 2 <= 1 > 0 < 9 !== 8 === 7 != 6 == 5 & 4 ^ 3 | 2 && 1 || 0 1`] = `
"SyntaxError [1:100-1:102]: Coalescing and logical operators used together in the same expression must be disambiguated with parentheses
> 1 | 3 ?? 2 ** 1 % 0 / 9 * 8 - 7 + 6 >>> 5 >> 4 << 3 >= 2 <= 1 > 0 < 9 !== 8 === 7 != 6 == 5 & 4 ^ 3 | 2 && 1 || 0
    |                                                                                                     ^^ Coalescing and logical operators used together in the same expression must be disambiguated with parentheses"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > c && d ?? e 1`] = `
"SyntaxError [1:7-1:9]: Coalescing and logical operators used together in the same expression must be disambiguated with parentheses
> 1 | c && d ?? e
    |        ^^ Coalescing and logical operators used together in the same expression must be disambiguated with parentheses"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > e ?? f ?? g || h; 1`] = `
"SyntaxError [1:12-1:14]: Coalescing and logical operators used together in the same expression must be disambiguated with parentheses
> 1 | e ?? f ?? g || h;
    |             ^^ Coalescing and logical operators used together in the same expression must be disambiguated with parentheses"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > function *f(){   s = {"foo": yield / x}   } 1`] = `
"SyntaxError [1:35-1:43]: Unterminated regular expression
> 1 | function *f(){   s = {"foo": yield / x}   }
    |                                    ^^^^^^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > function *f(){   s = {foo: yield / x}   } 1`] = `
"SyntaxError [1:33-1:41]: Unterminated regular expression
> 1 | function *f(){   s = {foo: yield / x}   }
    |                                  ^^^^^^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > function l(){((/)/))(/]/)}; 1`] = `
"SyntaxError [1:15-1:18]: Unterminated regular expression
> 1 | function l(){((/)/))(/]/)};
    |                ^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > function* f(){ yield
/foo } 1`] = `
"SyntaxError [2:0-2:6]: Unterminated regular expression
  1 | function* f(){ yield
> 2 | /foo }
    | ^^^^^^ Unterminated regular expression"
`;

exports[`Miscellaneous - Regular expressions > Miscellaneous - Regular expressions (fail) > s = {foo: yield / x} 1`] = `
"SyntaxError [1:10-1:15]: 'yield' may not be used as an identifier in this context
> 1 | s = {foo: yield / x}
    |           ^^^^^ 'yield' may not be used as an identifier in this context"
`;
